### ISSUE TITLE
Declare axiom compatibility and axiom classification

### ISSUE THREAD

**Issue by MSoegtropIMC** [coq/coq#10617]
This is a follow up on (https://coq.discourse.group/t/what-do-you-think-of-axioms/231/11)

The purpose is to make choices of axioms and axiom tracking more understandable to non logicians by declaring known information about well known or users axioms in order to produce warnings or errors if incompatible axioms are used and to make the output of `Print Assumptions` easier to read.

The idea is to have the following commands:
- `Declare Axioms Incompatible name1 name2` states that axioms name1 and name2 are incompatible. An error should be generated if axioms name1 and name2 (and this declaration) are loaded simultaneously. This error should be convertible to a warning with command line options (for people having issues with libraries using incompatible axioms but want to do some experiments)
- `Declare Axiom CoqCompatible name` for axioms which are (individually) well known to be compatible with the logic of Coq. `Print Assumptions` should have an option to ignore such axioms and also allow to print such classifications.

A few other axiom classifications are imaginable - possibly even user defined ones. Say:
```
Declare Axiom Class CoqCompatible. (* Axioms well known to be individually compatible with CIC *)
Declare Axiom Class Temporary. (* For things which we plan to prove soon *)
Declare Axiom Class Omission. (* For things which we expect to be able to prove but won't right now *)
Declare Axiom Class Controversial. (* For things we are not so sure about *)
Declare Axiom Class Reals. (* For Real number axioms *)
...
```
The main purpose of this is to structure the output of `Print Assumptions` so that controversial stuff gets higher visibility. Possibly one could add a numerical severity to the declaration, by which Print Assumptions sorts Axioms (it should anyway group by class).

**Comment by Zimmi48** [comment]
This is a nice idea but not realistic as is. Indeed, having a command that declares some relation like this between two axioms would need having both of the modules in which they are declared to be loaded (`Require`) which would already make the state inconsistent (in particular allow an automatic tactic to prove any goal by proving False).

Nonetheless, I see at least one partial solution to this which is to declare incompatibilities between libraries in the Coq package index, using opam's incompatible package feature. In order to make this more fine-grained, we could encourage library authors to structure their libraries with a single file declaring axioms that is loaded in all the other modules. If some modules do not depend on the axioms, we could encourage splitting the library in two separate opam packages.

As for adding annotations to single axioms (as opposed to pairs), this could be done with some new attributes.

**Comment by MSoegtropIMC** [comment]
> Indeed, having a command that declares some relation like this between two axioms would need having both of the modules in which they are declared to be loaded

Since this is more of an informative thing than anything hard I would say it is OK if one of the names is stored as string without any validity checking. Of cause this means that one doesn't find name changes during compilation, but one can have test cases which actually load two incompatible modules.

**Comment by JasonGross** [comment]
Perhaps the incompatibility could go through coqlib, so you don't need either to be loaded to declare them incompatible?  You'd need to be quite conservative in library structuring, though; I don't want loading `Program` to prevent me from using univalence because `Program` can use UIP.

**Comment by MSoegtropIMC** [comment]
> Perhaps the incompatibility could go through coqlib, so you don't need either to be loaded to declare them incompatible?

Could you elaborate on this idea a bit? I don't see how this would work.

**Comment by JasonGross** [comment]
You could do
```coq
Axiom UIP : ...
Register UIP as core.axiom.uip.
Declare Incompatible Axioms core.axiom.uip core.axiom.univalence.
```
in one file, and then in another write
```coq
Axiom univalence : ...
Register univalence as core.axiom.univalence.
Declare Incompatible Axioms core.axiom.uip core.axiom.univalence.
```
and then if you wrote
```coq
Assert Compatible Assumptions foo.
```
then Coq would error if `foo` depended on both something registered as `core.axiom.univalence` and something registered as `core.axiom.uip`.

**Comment by JasonGross** [comment]
This is quite similar to your "store as string" suggestion.

**Comment by MSoegtropIMC** [comment]
> ```coq
> Assert Compatible Assumptions foo.
> ```

I agree that for advanced users, your proposal of an explicit check is preferable because this simplifies library management in complex scenarios substantially.

But I also think that by default users should at least get a warning if they load modules containing incompatible axioms, so that they are aware. The warning can be switched off by command line and should point the user to some manual entry where the option of using explicit checks is explained.

**Comment by Zimmi48** [comment]
Using coqlib like you explained seems like an excellent idea @JasonGross!

Regarding warning vs `Assert`, I would favor the warning, because advanced users can turn this warning into an error.

I'll label this as "good first issue" because it seems that this issue contains the required technical hints to make this work. However, beware that there is still a significant design space regarding how to document the recommended use of this feature. It seems to me that a good solution would be to piggy-back on the existing wiki page https://github.com/coq/coq/wiki/CoqAndAxioms to add the official coqlib name to the page so that library authors start registering them. There should also be a call to library authors to add missing axioms to this page. Eventually, this page would make it to the refactored reference manual (more on this in an upcoming CEP).
